File: Example/Tests/SalesTests.m
//
//  TestDomainDataServices.m
//  WDePOS
//
//  Created by Danko, Radoslav on 17/10/15.
//  Copyright © 2015 Wirecard. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <libextobjc/EXTScope.h>
#import "UserHelper.h"
#import "SaleHelper.h"
#import "BaseTests.h"

@interface SalesTestsObjC : BaseTestsObcj
{

}

@end

@implementation SalesTestsObjC

#pragma mark - Helpers

- (void)setUp
{
    [super setUp];
}

-(void)testSales
{
    //PART 1: We log-in and request user data
    //--------------------------------------
    

    if (!loggedUser || ![loggedUser isKindOfClass:[WDMerchantUser class]])
    {
        XCTFail(@"Error, did not return Merchant User. Are your credentials correct? Try login into the backend through internet browser.");
    }
    
    
    //PART 2: We request sales from current user
    //----------------------------------------
    expectation = [self expectationWithDescription:@"Requesting Sales"];
    [self getSales];
    [self waitForExpectationsWithTimeout:100 handler:nil];
    if (!saleResponse)
    {
        XCTFail(@"Error getting a complete sale. Please create some test (ie. run CashTest). Otherwise, debug and check the returned error");
    }
    else
    {
        //We save one valid sale id for other tests cases such as Printer ones
        [[SaleHelper sharedInstance] saleToSaveIdFrom:saleResponse];
    }

}

-(void)getSales
{
    WDSalesQuery *query = [[WDSalesQuery alloc] init];
    query.page =  0; //Increase this if you want to do dynamic paging in your UI
    query.pageSize = 20; //Each sale includes a lot of data, so paging with small pageSizes is recommended
    query.orderBy = WDSaleQueryOrderByCreatedAt; //You can also sort by amount using WDTransactionsQueryOrderByAmount
    query.orderSort = WDQuerySortDescending;//Can be WDTransactionsQuerySortAscending
    query.statuses = @[@(WDSaleStateCompleted),
                       @(WDSaleStateFailed),
                       @(WDSaleStateReturned),
                       @(WDSaleStateCanceled)];//You can define more or less statuses for the query, depending on your needs.
    //You can use a receiptId provided by backend in query.receiptId if you have one
    [[sdk saleManager] querySales:query
                       completion:^(NSArray<WDSaleResponse *> * arr, NSError * error)
    {
        saleResponse = [arr firstObject];
        returnedErr = error;
        [expectation fulfill];
    }];
}


@end

File: Example/Tests/SalesTests.swift
//
//  UserTestsSwift.swift
//  DemoEPOS
//
//  Created by Fortes, Francisco on 7/25/17.
//  Copyright © 2017 Wirecard. All rights reserved.
//

import Foundation
import XCTest
import WDePOS

class SalesTestsSwift: BaseTestsSwift
{
    override func setUp()
    {
        super.setUp()
    }
    func testSales()
    {
        //PART 1: We log-in and request user data
        //--------------------------------------
       

        if (loggedUser == nil || (loggedUser?.isKind(of: WDMerchantUser.self)) == false)
        {
            XCTFail("Error, did not return Merchant User. Are your credentials correct? Try login into the backend through internet browser.")
        }
        
        //PART 2: We request sales from current user
        //----------------------------------------
        expectation = self.expectation(description: "Requesting Sales")
        self.getSales()
        self.waitForExpectations(timeout: 100, handler: nil)
        if (self.saleResponse == nil)
        {
            XCTFail("Error getting a complete sale. Please create some test (ie. run CashTest). Otherwise, debug and check the returned error.")
        }
        else
        {
            SaleHelper.sharedInstance().saleToSaveId(from: self.saleResponse)
        }
    }
    
    func getSales()
    {
        let statuses = [NSNumber].init(arrayLiteral: NSNumber.init(value: WDSaleState.completed.rawValue),
                                       NSNumber.init(value:WDSaleState.failed.rawValue),
                                       NSNumber.init(value: WDSaleState.returned.rawValue),
                                       NSNumber.init(value: WDSaleState.canceled.rawValue))
        let types = [NSNumber].init(arrayLiteral:NSNumber.init(value: WDSaleType.purchase.rawValue))        
        let query : WDSalesQuery = WDSalesQuery.init(page: 0,
                                                                 pageSize: 20,
                                                                 orderBy: .createdAt,
                                                                 order: .descending,
                                                                 statuses: statuses,
                                                                 saleTypes: types)
        sdk.saleManager.querySales(query, completion: {[weak self](arr : [WDSaleResponse]?, error: Error?) in
            self?.saleResponse = arr?.first
            self?.returnedErr = error
            self?.expectation.fulfill()
        })
    }
}
